pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -735,14 +742,45 @@ protected SegmentBuildDescriptor buildSegmentInternal(boolean forCommit) {
     if (_isOffHeap) {
       params.withMemoryUsedBytes(_memoryManager.getTotalAllocatedBytes());
     }
-    SegmentCompletionProtocol.Response commitEndResponse = _protocolHandler.segmentCommitEnd(params);
+    SegmentCompletionProtocol.Response commitEndResponse;
+    if (!_indexLoadingConfig.isEnableSplitCommitEndWithMetadata()) {
+      commitEndResponse = _protocolHandler.segmentCommitEnd(params);
+    } else {
+      Map<String, File> metadataFiles = new HashMap<>();
+      try {
+        metadataFiles.put(V1Constants.MetadataKeys.METADATA_FILE_NAME,
```

Comment:
Can you find a way to avoid extracting from the tar file again? CPU and disk IO are expensive on the server, and we would like to avoid it if we can. 

The only way I could think of was to store the metadata files in another directory and keep it along with the segment descriptor. Once the commit is completed, we need to take care to remove the directory, as we don't want to have cruft left-over. It has to be done a little carefully. @npawar  do you have other ideas?

Reply:
Let me explore a bit your idea to pass in the metadata files via the Segment Descriptor.

Reply:
I think we keep the raw files and build the tar.gz in addition. The segment needs to be in raw format for serving anyways.

Reply:
I see. Then we can just use the existing metadata files on the server -- and there is no need to delete any temp files. I tested it and it works. Please check the latest commits.